[RFC] Expand cache management APIs

carlocaione opened the pull request:
This is my personal shot at #2647. Admittedly my goal was not creating a full fledged cache management framework but I tried to add the missing pieces that I need for my work on AArch64 that could be used by other architectures.

carlocaione committed:
cache: Expand the APIs for cache flushing

The only two supported operations for data caches in the cache framework
are currently arch_dcache_flush() and arch_dcache_invd().

This is quite restrictive because for some architectures we also want to
control i-cache and in general we want a finer control over what can be
flushed, invalidated or cleaned. To address these needs this patch
expands the set of operations that can be performed on data and
instruction caches, adding hooks for the operations on the whole cache,
a specific level or a specific address range.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
kconfig: Unify CACHE_* options

The kconfig options to configure the cache flushing framework are
currently living in the arch-specific kconfigs of ARC and X86 (32-bit)
architectures even though these are defining the same things.

Move the common symbols in one place accessible by all the architectures
and create a menu for those.

Leave the default values in the arch-specific locations.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arc: cache: Use new cache APIs

Add an helper to correctly use the new cache APIs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
x86: cache: Use new cache APIs

Add an helper to correctly use the new cache APIs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
cache: libmetal: Redirect sys_cache_flush()

Libmetal is still using the old sys_cache_flush(). Use an helper until
libmetal is properly fixed.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

abrodkin commented on arch/arc/core/cache.c:
```diff
@@ -25,6 +25,10 @@
 #include <init.h>
 #include <stdbool.h>
 
+#if defined(CONFIG_CACHE_LINE_SIZE_DETECT)
```
Why moving this from already existing ifdef section for `CONFIG_CACHE_LINE_SIZE_DETECT` below?
IMHO less ifdefs makes code simpler and easier to digest.

carlocaione replied:
Because otherwise it fails to compile when enabling `ARCH_CACHE_FLUSH_DETECT`

abrodkin reviewed (dismissed):
Thanks for doing that hard work on generalization of commonly used things!
See my comments though below.

carlocaione commented on arch/Kconfig:
```diff
@@ -758,6 +758,34 @@ config FPU_SHARING
+
+config CACHE_LINE_SIZE_DETECT
+	bool "Detect cache line size at runtime"
+	default y
```
Agree, no problem in switching that to a default `n`.

carlocaione commented on arch/Kconfig:
```diff
@@ -758,6 +758,34 @@ config FPU_SHARING
+config CACHE_FLUSHING
+	bool "Enable cache flushing mechanism"
+	help
+	  This links in the cache flushing functions (d-cache and i-cache).
```
I see how the i-cache operations can be a bit atypical on some architectures but right now the `CACHE_FLUSHING` is an umbrella for both d-cache and i-cache operations.

carlocaione commented on arch/Kconfig:
```diff
@@ -758,6 +758,34 @@ config FPU_SHARING
+	bool "Detect cache line size at runtime"
+	default y
+	help
+	  This option enables querying the d-cache register for finding the
```
ACK

carlocaione commented on include/cache.h:
```diff
@@ -13,57 +13,178 @@
+#define K_CACHE_INVD		BIT(1)
+#define K_CACHE_FLUSH		BIT(2)
+#define K_CACHE_FLUSH_INVD	(K_CACHE_FLUSH | K_CACHE_INVD)
+#define K_CACHE_CLEAN		BIT(3)
```
At least on AArch64 you can have the three operations: flush (`dc civac`), clean (`dc cvac`) and invalidate (`dc ivac`) so I'm trying to consider all the possible cases here.

iMitsyanko replied:
Hi @carlocaione , could you please confirm what is a "clean" operation meaning? Looking through ARM docs, it seems "clean" is an ARM term for "flushing" (which is used by ARC, for example), so clean==flush.
Regarding instructions that you listed, ARMv8 ARM defines them as:
DC CVAC - Clean by virtual Address
DC CIVAC - Clean and invalidate by VA
DC IVAC - Invalidate by VA

carlocaione replied:
Hi @iMitsyanko. AFAIK in the AArch64 world usually the `flush` operation is `clean + invalidate`, see for example https://github.com/ARM-software/arm-trusted-firmware/blob/master/lib/aarch64/cache_helpers.S#L43-L44 or https://github.com/u-boot/u-boot/blob/master/arch/arm/cpu/armv8/cache.S#L132-L142 while the `clean` operation without `invalidate` is a different operation.

So, for AArch64 we have:
- `clean`
- `invalidate`
- `flush == clean + invalidate`

I decided to keep flush as standalone operation because I don't know how these are mapped on other architectures so I tried to be as general as possible with `clean`, `invalidate` and `flush` as three distinct operations. Each arch can map the three ops as they please.

iMitsyanko replied:
This definitely makes sense, though it seems in that case K_CACHE_FLUSH_INVD and K_CACHE_CLEAN_INVD definitions should be removed from this PR?

carlocaione replied:
No objection to remove those. I can leave only `K_CACHE_INVD` `K_CACHE_FLUSH` and `K_CACHE_CLEAN` and each arch can `OR` them if they want to.

sandeepbrcm replied:
One thing to note : Most of the latest armv8 cores do not really differentiate the clean and invalidate CMOs. ie. DC IVAC (invalidate) is promoted to DC CIVAC (clean and invalidate).

From OS perspective if we have architecture agnostic range based flush for I$ D$ that should be sufficient. Do you have a use case in mind for the granular `clean` operation. How it will be used ?
Rest of the APIs looks good to me.

carlocaione replied:
I don't have in mind any use case but we cannot exclude that some architecture can actually fully differentiate between `flush` `clean` and `invalidate`, so I'm just trying to be a bit fool / future proof here leaving all the options on the table.

pizi-nordic replied:
@carlocaine IMHO we should not use ARM nomenclature (`flush`/`invalidate`/`clean`) which might be misleading on non-ARM machines, but instead name the operations basing on what they do: `write_back`, `invalidate` and `write_back_invalidate`.

carlocaione replied:
Interesting. Is that really ARM nomenclature? At least ARC, X86 and NIOS2 seem to use `flush` with a similar meaning. I'm not strongly opinionated on the naming so I'll change it for the sake of moving this forward (but admittedly I like `flush` more than `write_back`)

pizi-nordic replied:
Note that the proposed names can be shorted to wb/inv/wb_inv to make lazy programmers happy :)

carlocaione replied:
updated ;)

carlocaione commented on include/cache.h:
```diff
@@ -13,57 +13,178 @@
+__syscall int sys_dcache_level(int level, int op);
+static inline int z_impl_sys_dcache_level(int level, int op)
+{
+	if (IS_ENABLED(CONFIG_CACHE_FLUSHING)) {
```
Yes, you are totally right that `CACHE_FLUSHING` is misleading here. I'll change that.

carlocaione commented on arch/arc/core/cache.c:
```diff
@@ -126,8 +130,20 @@ void arch_dcache_invd(void *start_addr_ptr, size_t size)
+	} else if (op == K_CACHE_FLUSH) {
+		arch_dcache_flush(addr, size);
+	} else {
+		return -ENOTSUP;
```
This is was a guess on my side because I'm not familiar with the ARC architecture. I'll try to fix that as best as I can.

carlocaione commented on arch/arc/core/cache.c:
```diff
@@ -67,7 +71,7 @@ static void dcache_enable(void)
 }
 
-void arch_dcache_flush(void *start_addr_ptr, size_t size)
+static void arch_dcache_flush(void *start_addr_ptr, size_t size)
```
Right, I'll fix that.

carlocaione committed:
cache: Rename CACHE_FLUSHING to CACHE_MANAGEMENT

The new APIs are not only dealing with cache flushing. Rename the
Kconfig symbol to CACHE_MANAGEMENT to better reflect this change.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

abrodkin commented:
@evgeniy-paltsev could you please also take a look at this one?

carlocaione commented:
Any more feedback on this? @abrodkin @evgeniy-paltsev

martin-aberg commented on include/cache.h:
```diff
@@ -13,59 +13,185 @@
 
-void arch_dcache_flush(void *addr, size_t size);
-void arch_dcache_invd(void *addr, size_t size);
+/**
+ * Common operations for the caches
+ */
+#define K_CACHE_WB	BIT(0)
+#define K_CACHE_INVD	BIT(1)
+#define K_CACHE_WB_INVD	(K_CACHE_WB | K_CACHE_INVD)
```
Could add a few word describing the intention of the operations here, or make reference to a description. Previous discussions in this PR has identified that there might be different interpretations and terminology depending on context.
Something along the lines of
* _"WB means write-back and intends to transfer dirty cache lines to memory in a copy-back cache policy. May be a no-op in write-back cache policy."_
* _"INVD means invalidate and will mark cache lines as not valid. A future access to the associated address is guaranteed to generate a memory fetch."_

carlocaione replied:
ACK

martin-aberg commented on include/cache.h:
```diff
@@ -13,59 +13,185 @@
+ * @param op	Operation to perform (one of the K_CACHE_* operations)
+ *
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
```
Does _"the operation is not supported"_ include the case when `K_CACHE_WB_INVD` is attempted on a cache which supports the `INVD` but not the `WB`? In other words, if the application needs to know the cache capabilities.

carlocaione replied:
I didn't attach any particular meaning to that but your interpretation matches mine. If the arch returns `-ENOTSUP` that means that that particular operation is not supported so yes, the application can use that to probe the cache capabilities.

martin-aberg commented on include/cache.h:
```diff
@@ -13,59 +13,185 @@
+ * @retval -ENOTSUP	If the operation is not supported
  */
-__syscall void sys_cache_flush(void *addr, size_t size);
+int arch_dcache_range(void *addr, size_t size, int op);
```
What would be the correct interpretation of this function on a CPU cache controller which can perform an operation (for example `INVD`) on the full cache but not on a per-line basis? (The LEON3 processors is one such example.) I see two main interpretations:
1. Do the operation on the full i-cache or d-cache. `return 0`.
or
2. Do no operation. `return -ENOTSUP`.

carlocaione replied:
I'd say that this operation would return `-ENOTSUP` but you can still use `arch_dcache_all / arch_icache_all` in that case (and that should return `0`).

martin-aberg replied:
Thanks for the clarification. It makes sense for the low-level API.

One use case I had in mind was invalidating/flushing buffers in DMA based peripheral drivers on systems which do not have cache coherency between CPU and DMA. If we consider a scenario where the driver wants to invalidate cache state for 10 small buffers located at various places in RAM because a DMA read operation has finished. Then if the CPU has the capability to operate per-line, then it is probably best to do 10 individual `arch_dcache_range()` in a loop over DMA buffer desriptors. On a CPU which lacks the per-line invalidation capability, the right thing to do is likely a single invalidation with `arch_dcache_all()`.

That use case is all fine with the suggested API.

martin-aberg reviewed (commented):
Thank you @carlocaione for the work on the API.
I have added some comments from the point of implementing the API on the `arch` level.

carlocaione commented on arch/Kconfig:
```diff
@@ -717,6 +717,34 @@ config FPU_SHARING
+	  If the CPU's cache line size is known in advance, disable this option and
+	  manually enter the value for CACHE_LINE_SIZE.
+
+config CACHE_LINE_SIZE
+	int "Cache line size" if !CACHE_LINE_SIZE_DETECT
```
DT defines `i-cache-size`, `d-cache-size` and `cache-size` for the `cpu` nodes. I can add an helper function to get that from the DTS but that would require some changes to the current X86 and ARC code.

carlocaione commented on arch/Kconfig:
```diff
@@ -717,6 +717,34 @@ config FPU_SHARING
+	int "Cache line size" if !CACHE_LINE_SIZE_DETECT
+	default 0
+	help
+	  Size in bytes of a CPU cache line.
+
+	  Detect automatically at runtime by selecting CACHE_LINE_SIZE_DETECT.
```
It is indeed. I wanted to keep this a bit easier than it is turning out to be ;) At this point the solution I see is:
1. Add two Kconfig symbols `DCACHE_LINE_SIZE` and `ICACHE_LINE_SIZE`
2. Add two helper functions `arch_dcache_line_size()` and `arch_icache_line_size()` that read the values from DT if the Kconfig symbols are not defined
3. ?
4. Profit

carlocaione commented:
@martin-aberg I pushed a new version addressing your suggestions. The biggest change is that now the d-cache and i-cache line size can be obtained in 3 ways (in order): 
1. The arch can define an hw-dependent way to get that by chosing `CONFIG_DCACHE_LINE_SIZE_DETECT` and `CONFIG_ICACHE_LINE_SIZE_DETECT` and defining the arch-dependent functions `arch_dcache_line_size_get` and `arch_icache_line_size_get`
2. By using the Kconfig symbols `CONFIG_DCACHE_LINE_SIZE` and `CONFIG_ICACHE_LINE_SIZE`
3. Adding the `d-cache-line-size` and `i-cache-line-size` properties to the DT

carlescufi commented:
@abrodkin please re-review

evgeniy-paltsev commented on include/cache.h:
```diff
@@ -13,78 +13,229 @@
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
+ */
+int arch_dcache_level(int level, int op);
```
Hmm, I'm wondering do we really need to expand such low-level cache management API to generic code?
Any example about why generic code may want to do some operations with exact cache level?

carlocaione replied:
reading around it seems that sometimes for example you want to invalidate L1 cache on ARM v7 cores but it's really a narrow use case and for such cases it's probably better to just add a few lines of code when needed. So I agree that it's better to take this out.

abrodkin reviewed (approved):
We seem to not have instruction cache callbacks at the moment for ARC, but that's purely our own problem which we'll fix later if needed. But otherwise for now all looks good. And thanks for doing all that work for all of us!
